clr: Implement dynamic stream to HWq logic

saleelk opened the pull request:
* This change implements a queue depth heuristics with weights to get
  optimal HWq assignment to a stream
* Use DEBUG_HIP_DYNAMIC_QUEUE for dynamic queues.
  0 - disabled, 1 - Queue Depth heuristics 2 - Queue+pipe_dist(Default)
* Use env DEBUG_HIP_IGNORE_STREAM_PRIORITY=1 to ignore priority stream
  creation

## Motivation
Better stream to queue mapping

## Technical Details

## Test Plan
PSDB tests, Aorta
## Test Result
Improves performance significantly in certain cases

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

gandryey commented on projects/clr/rocclr/device/rocm/rocvirtual.cpp:
```diff
@@ -2033,8 +2069,19 @@ void VirtualGPU::ReleaseHwQueue() {
+  // Dedicated queues keep their HW queue, never acquire from pool
+  if (!dedicated_queue_ && gpu_queue_ == nullptr) {
+    gpu_queue_ = roc_device_.AcquireActiveQueue(priority_, last_assigned_queue_);
+    if (gpu_queue_ != nullptr) {
```
Acquire can't return null queue. We will crash

saleelk replied:
ACK

gandryey commented on projects/clr/rocclr/device/rocm/rocvirtual.cpp:
```diff
@@ -4051,8 +4098,11 @@ void VirtualGPU::submitMarker(amd::Marker& vcmd) {
-        gpu_queue_ = roc_device_.AcquireActiveNormalQueue();
+      if (!dedicated_queue_ && gpu_queue_ == nullptr) {
+        gpu_queue_ = roc_device_.AcquireActiveQueue(priority_, last_assigned_queue_);
+        if (gpu_queue_ != nullptr) {
```
Acquire can't return null queue. We will crash

saleelk replied:
ACK

gandryey commented on projects/clr/rocclr/device/rocm/rocvirtual.hpp:
```diff
@@ -539,24 +544,43 @@ class VirtualGPU : public device::VirtualDevice {
+    uint64_t wptr = Hsa::queue_load_write_index_relaxed(gpu_queue_);
+    uint64_t rptr = Hsa::queue_load_read_index_relaxed(gpu_queue_);
+
+    if (wptr == rptr) {
```
According to the AQL spec the state of read/write pointer can't be used as the check for the queue idle state.

saleelk replied:
ACK

saleelk committed:
clr: Implement dynamic stream to HW queue assignment

This change implements dynamic stream to hardware queue (HWq) mapping
with the following features:

* Queue depth heuristics with weights for optimal HWq assignment
* Make last used queue sticky for better locality
* Use pipe HWq to pipe mapping - gfx9 follows a round-robin queue to
  pipe mapping based on creation order (single process per device only,
  as pipe ID is statically assigned by runtime)
* More aggressive heuristic usage for better queue distribution
* Extend dynamic queues support for all stream priorities

Environment variables:
* DEBUG_HIP_DYNAMIC_QUEUE: 0 - disabled, 1 - Depth heuristics 2 -
  Depth+Pipe heuristics
* DEBUG_HIP_IGNORE_STREAM_PRIORITY=1: ignore priority stream creation

gandryey commented on projects/clr/rocclr/device/rocm/rocvirtual.hpp:
```diff
@@ -539,24 +544,38 @@ class VirtualGPU : public device::VirtualDevice {
+  }
+
+  //! Get the current queue depth (number of pending packets)
+  uint64_t GetQueueDepth() const {
```
It's not used anywhere

gandryey commented on projects/clr/rocclr/device/rocm/rocvirtual.cpp:
```diff
@@ -1986,26 +1998,39 @@ address VirtualGPU::allocKernelArguments(size_t size, size_t alignment) {
+  }
+
+  // Try to release queue to the pool of active queues
+  if (roc_device_.settings().dynamic_queues_ >= 1 && !cooperative_ &&
```
Could you replace with > 0 or keep just roc_device_.settings().dynamic_queues_?

saleelk committed:
clr: Clean up last_used_queue_
